core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -556,6 +557,87 @@ public static boolean isExactNumeric(RelDataType type) {
     }
   }
 
+  /**
+   * Returns whether {@code container} can represent every value produced by
+   * {@code content} without loss of information.
+   *
+   * <p>The {@code container} type must be one of the integer types (signed or
+   * unsigned). The {@code content} type can be integer, or a DECIMAL with
+   * scale {@code 0}. For all other types this method returns {@code false}.
+   *
+   * @throws IllegalArgumentException if {@code container} is not an integer type
+   */
+  @API(since = "1.40", status = API.Status.EXPERIMENTAL)
```

Comment:
If the `isLosslessCast` method has been marked as stable, is it necessary to mark this method as experimental?

Reply:
That's an interesting point, we probably want to keep both experimental or stable.

To me experimental is really about breaking the contract without waiting for a new major version, but here the only relevant contrat is that the `isLosslessCast` method is sound but not necessarily complete, and we will ever want to make it unsound on purpose.

The new method seems reasonably complete and I don't see it changing existing behavior unless we find a bug.

All considered, we could promote both to Stable, WDYT?